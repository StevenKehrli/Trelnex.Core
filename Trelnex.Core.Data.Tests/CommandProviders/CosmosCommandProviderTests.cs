using Azure.Identity;
using Microsoft.Azure.Cosmos;
using Microsoft.Azure.Cosmos.Linq;
using Microsoft.Extensions.Configuration;

namespace Trelnex.Core.Data.Tests.CommandProviders;

[Ignore("Requires a CosmosDB instance.")]
public class CosmosCommandProviderTests : CommandProviderTests
{
    private Container _container = null!;

    [OneTimeSetUp]
    public async Task TestFixtureSetup()
    {
        // This method is called once prior to executing any of the tests in the fixture.

        // create the test configuration
        var configuration = new ConfigurationBuilder()
            .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
            .AddJsonFile("appsettings.User.json", optional: true, reloadOnChange: true)
            .Build();

        var providerConfiguration = configuration.GetSection("CosmosCommandProviders").Get<CosmosCommandProviderConfiguration>()!;

        // create a cosmos client for cleanup
        var tokenCredential = new DefaultAzureCredential();

        var cosmosClient = new CosmosClient(
            accountEndpoint: providerConfiguration.EndpointUri,
            tokenCredential: tokenCredential);

        _container = cosmosClient.GetContainer(
            databaseId: providerConfiguration.DatabaseId,
            containerId: providerConfiguration.ContainerId);

        // create the command provider
        var cosmosClientOptions = new CosmosClientOptions(
            TokenCredential: tokenCredential,
            AccountEndpoint: providerConfiguration.EndpointUri,
            DatabaseId: providerConfiguration.DatabaseId,
            ContainerIds: [ providerConfiguration.ContainerId ]
        );

        var keyResolverOptions = new KeyResolverOptions(
            TokenCredential: tokenCredential);

        var factory = await CosmosCommandProviderFactory.Create(
            cosmosClientOptions,
            keyResolverOptions);

        _commandProvider = factory.Create<ITestItem, TestItem>(
            providerConfiguration.ContainerId,
            "test-item",
            TestItem.Validator,
            CommandOperations.All);
    }

    [TearDown]
    public async Task Cleanup()
    {
        // This method is called after each test case is run.

        var feedIterator = _container
            .GetItemLinqQueryable<CosmosItem>()
            .ToFeedIterator();

        while (feedIterator.HasMoreResults)
        {
            var feedResponse = await feedIterator.ReadNextAsync();

            foreach (var item in feedResponse)
            {
                await _container.DeleteItemAsync<CosmosItem>(
                    id: item.id,
                    partitionKey: new PartitionKey(item.partitionKey));
            }
        }
    }

    /// <summary>
    /// Represents the configuration properties for Cosmos command providers.
    /// </summary>
    private record CosmosCommandProviderConfiguration(
        string EndpointUri,
        string DatabaseId,
        string ContainerId);

    private record CosmosItem(
        string id,
        string partitionKey);
}
