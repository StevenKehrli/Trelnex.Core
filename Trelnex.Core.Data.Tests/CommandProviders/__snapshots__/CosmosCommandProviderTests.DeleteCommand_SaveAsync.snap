{
  "Item": {
    "Id": "f8829dac-56f6-4448-829a-fac886aefb1b",
    "PartitionKey": "a3e68ca8-2555-4f03-b58f-c0e80bcd38f1",
    "TypeName": "test-item",
    "CreatedDate": "2025-01-03T03:11:00.745459Z",
    "UpdatedDate": "2025-01-03T03:11:00.745459Z",
    "DeletedDate": "2025-01-03T03:11:02.50113Z",
    "IsDeleted": true,
    "ETag": "\"6301c5b4-0000-0300-0000-677755460000\"",
    "Message": "Message #1"
  }
}
